Add ability to disable schema-json and graphiql endpoints

bono007 opened the pull request:
@paulbakker @berngp @marceloverdijk

This is "Part I" as described in the issue. 

This proposal is still missing (I will do this later today):
*  properties in additional meta info file
* docs

Fixes gh-88

bono007 commented on graphql-dgs-graphiql-autoconfigure/src/test/kotlin/com/netflix/graphql/dgs/graphiql/autoconfiguration/GraphiQLAutoConfigurationTest.kt:
```diff
@@ -0,0 +1,48 @@
+import org.springframework.boot.autoconfigure.AutoConfigurations
+import org.springframework.boot.test.context.runner.WebApplicationContextRunner
+
+class GraphiQLAutoConfigurationTest {
```
I know there was no tests in the auto-config modules for graphiql or mvc but I think its worth putting the tests in here rather than rely on them being in the "outer" autoconfig module as its easier for future contributors to see where the auto-config tests are and as this grows bigger so will these tests.

berngp commented on graphql-dgs-spring-webmvc/src/main/kotlin/com/netflix/graphql/dgs/mvc/DgsRestController.kt:
```diff
@@ -64,10 +61,10 @@ import org.springframework.web.multipart.MultipartFile
     val logger: Logger = LoggerFactory.getLogger(DgsRestController::class.java)
 
-    @RequestMapping("/graphql", produces = ["application/json"])
+    @RequestMapping("\${dgs.graphql.path:/graphql}", produces = ["application/json"])
```
I like detaching the definition of default values from the place where they are used. This can be done with an EnvironmentPostProcessor that loads a _default properties file_. 

That said, I think this should be considered as a separate pr.

bono007 replied:
Thats a good idea especially when the number of properties grows and you can see all of the defaults in a single place. 

The one area where this is, not in conflict, but would need to be agreed upon is the use of `@DefaultValue` in the `@ConfigurationProperties`. It would be confusing having defaults in the config props and the defaults file. We could just make a guideline to not use `@DefaultValue` in config props though.

berngp reviewed (approved):
yhank you form the contribution and specially for the arguments you made on the issue that originated this change.

The only thing I would say is missing is the configuration metadata.

Red https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-configuration-metadata.html

bono007 commented:
> yhank you form the contribution and specially for the arguments you made on the issue that originated this change.
> 
> The only thing I would say is missing is the configuration metadata.
> 
> Red https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-configuration-metadata.html

You are more than welcome @berngp . I plan on doing docs and config props in this MR but later today as a follow on commit. I'll ping you when its in.

bono007 commented:
@berngp @paulbakker @marceloverdijk 

I pushed a 2nd commit that adds the config metadata files along w/ the `kapt` plugin to process them. This sets us up nice when we start adding `@ConfigurationProperties`, they should just work. 

I verified in one of my teams app that uses the starters and I get this beauty in the IDE....

<img width="856" alt="Screen Shot 2021-02-28 at 9 09 09 PM" src="https://user-images.githubusercontent.com/28907971/109448241-1fb33800-7a0b-11eb-96d2-c2c4914bd7dc.png">

I have added some inline comments to the review.

bono007 commented on buildSrc/src/main/kotlin/Versions.kt:
```diff
@@ -16,6 +16,7 @@
 
 object Versions {
     const val KOTLIN_VERSON = "1.4.10"
+	const val KAPT_VERSION = "1.4.10"
```
There is a more recent version but when I moved it and Kotlin version out of sync - it did not end well.

bono007 commented on graphql-dgs-graphiql-autoconfigure/build.gradle.kts:
```diff
@@ -16,9 +16,23 @@
+	testImplementation("org.springframework.boot:spring-boot-starter-web")
+}
+
+kapt {
```
This should not be necessary and is probably a bug in kapt or the boot config processor - trust me, it does not include the metadata w/o this - it robbed me of 1-2 hrs of time. It works w/ this though. We will probably want to factor this out for easier re-use when we start adding config props.

bono007 commented on graphql-dgs-graphiql-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json:
```diff
@@ -0,0 +1,10 @@
+    {
+      "name": "dgs.graphql.graphiql.enabled",
+      "type": "java.lang.Boolean",
+      "description": "Enables GraphiQL functionality.",
```
**[HELP_SOLICITED]** not the best description.

bono007 commented on graphql-dgs-spring-webmvc-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json:
```diff
@@ -0,0 +1,16 @@
+    {
+      "name": "dgs.graphql.schema-json.enabled",
+      "type": "java.lang.Boolean",
+      "description": "Enables schema-json endpoint functionality.",
```
**[HELP_SOLICITED]** not the best description.

bono007 commented:
I think we should add a CONTRIBUTING.md or some developer facing light doc that tells how to run, test, and now about additional config props, regular config props, and kapt. Not a manual, just some important lightweight instruction. 

Thoughts?

berngp commented:
Will review further tomorrow, some remarks...

I would like to avoid kapt and instead use Spring Boot's annotation post-processor for gradle, ref [here](https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-configuration-metadata.html#configuration-metadata-annotation-processor-setup).


Agree on all the points you mentioned about the guidelines. On fact I'm interested at defining auto-formatting via spotless (foe example)

And again, thanks for all this work!

bono007 commented:
> I would like to avoid kapt and instead use Spring Boot's annotation post-processor for gradle, ref here.

@berngp I believe you have to use `kapt` in order to process the `@ConfigurationProperties` - thats what I picked up from [this](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-kotlin-configuration-properties).

paulbakker commented on graphql-dgs-spring-webmvc-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "properties": [
```
Shouldn't `dgs.graphql.path` be included as well?

bono007 replied:
I split out the path config based on our earlier discussion.

paulbakker reviewed (commented):
> > I would like to avoid kapt and instead use Spring Boot's annotation post-processor for gradle, ref here.
> 
> @berngp I believe you have to use `kapt` in order to process the `@ConfigurationProperties` - thats what I picked up from [this](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-kotlin-configuration-properties).

The docs are a little confusing; is kapt only required to work with immutable config classes, or is it required to get it to work at all? I _think_ it's the latter, but I'm not sure.

paulbakker commented:
> > The docs are a little confusing; is kapt only required to work with immutable config classes, or is it required to get it to work at all? I _think_ it's the latter, but I'm not sure.
> 
> If that is the case, it will come down to a "What is worse, kapt or mutable config props?" decision. I don't have any ill feeling towards kapt yet and I have really grown to ❤️ my immutable config props. hehe. I am guessing that there are negatives and hence a desire to avoid kapt.

I have no experience with it, but generally prefer to not rely on too many plugins. However, this seems to be a very common one, so I have no issue with it.

paulbakker commented:
@bono007 Everything looks good to me - happy to merge whenever you're ready.

berngp commented:
@bono007 if you rebase against master you don't need to set `kapt` nor configure it to include any kind of resource.

bono007 commented:
Awesome! I saw the email notifications but have not got a chance to rebase.
Will do that this evening as well as split out the path configuring portion
into follow on mr.

On Mon, Mar 1, 2021, 5:55 PM Bernardo Gomez Palacio <
notifications@github.com> wrote:

> @bono007 <https://github.com/bono007> if you rebase against master you
> don't need to set kapt nor configure it to include any kind of resource.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Netflix/dgs-framework/pull/94#issuecomment-788410396>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG4RTQ42OB7M6TWO45DSLGDTBQSPRANCNFSM4YLHHFWA>
> .
>

bono007 commented on graphql-dgs-spring-webmvc-autoconfigure/src/test/kotlin/com/netflix/graphql/dgs/webmvc/autoconfiguration/DgsWebMvcAutoConfigurationTest.kt:
```diff
@@ -0,0 +1,128 @@
+
+    @Test
+    fun schemaJsonControllerMappedToDefaultPath() {
+        context.withConfiguration(AutoConfigurations.of(JacksonAutoConfiguration::class.java, WebMvcAutoConfiguration::class.java, DispatcherServletAutoConfiguration::class.java))
```
This and the following test are new since last commit. I did not have any coverage around switching the schema-json path and it actually service at that path. This adds a bit of Fu to the context runner to allow mockmvc in the runner. Its a trick I picked up from [Spring Boot](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/integrationtest/WebMvcEndpointCorsIntegrationTests.java#L51).

bono007 commented:
Ok @paulbakker @berngp .. I rebased and ended up squashing as it was getting confusing to look at from a commit history perspective. 

Here is the summary of the current bits:
- add ability to disable graphiql and schema-json endpoints
- add ability to configure schema-json path 
- all properties added in metadata files
- all functionality unit/integration tests 

Here is what I removed:
- ability to configure graphql path 

Here is what I will do next (after this is merged):
- start PR that configures graphql path (the portion I removed from above)
- this PR will eventually hold the changes to configure graphiql path as well

berngp commented:
Thanks for all the work with this PR @bono007 !
